man/bmail.8
.\" $Header$
.\"
.\"  Man page for HUJI-NJE/FUNET-NJE on UNIX system.
.\"
.\"  Written by  Matti Aarnio <mea@finfiles.bitnet> <mea@nic.funet.fi>
.\"
.\"  Date: 27-Dec-1990, 18-Sep-1993
.\"
.ds ]W Funet-Nje 1 UNIX 3.0
.TH BMAIL 8
.SH NAME
`NJELIB'/bmail \- bitnet mail send tool
.SH SYNOPSIS
.IP bmail 0.6i 0
.RB [ " \-nd Gateway-address " ]
.RB [ " \-b(smtp) Gateway-address " ]
.RB [ " \-tag RSCS-TAG " ]
.RB [ " \-u origuser " ]
.if n .br
.I FromUser@FromNode
.I ToUser@ToNode
.RI [ " ToUser@ToNode... " ]
.SH DESCRIPTION
.IX "bitnet"
.IX "funetnje"
.IX "bmail"
.LP
.I bmail
command is actually a special purpose \fIsendfile\fR
(Historically: code of \fIsendfile\fR reused a lot of \fIbmail\fR code)
for mail transport, and thus is not a general user program.
Its input is a ready-made mail file, and it generates a BSMTP envelope,
when so requested.
.LP
Options for
.I bmail
are:
.IP "\fB\-b(smtp) gateway@host\fR" 1.8i 0
informing that the feed in mail shall be wrapped into a BSMTP/PUNCH envelope
with "MAIL FROM:<>" getting the first argument, and "RCPT TO:<>" getting
the rest (1 to many..)
This option has two alternate names: "\fB-b\fR", and "\fB-bsmtp\fR".
.I Line length is limited to 80 chars on this mode!  Wrapping of longer lines
.I is attempted, but it might not always be successful.
.IP "\fB\-nd gateway@host\fR" 1.8i 0
informing that the feed in mail shall be wrapped into a BSMTP/NETDATA envelope
with "MAIL FROM:<>" getting the first argument, and "RCPT TO:<>" getting
the rest (1 to many..)
.I This accept indefinite length lines, minimum being 80 chars, true value
.I is calculated while preparing the BSMTP envelope into a temporary file.
.IP "\fB\-tag RSCS-TAG\fR" 1.8i 0
defines the 136 bytes (padded/truncated to that length) RSCS TAG data.
Usually not needed with
.IR bmail .
.IP "\fB\-u origuser\fR" 1.8i 0
Can be used to define local mail-sender pseudo-id.
For
.SM BSMTP
modes the default value is
.BR MAILER .
For others it is local user.
.LP
When
.IR bmail's
`From' argument contains a `.' in it, the file sender is changed to
be "MAILER@localname", however possible BSMTP will contain entered
`From' information.
.SH INTERFACING TO SYSTEM MAILER
In case of
.BR zmailer (8z)
by Rayan Zachariassen, interface goes as follows:
.LP
.nf
.ft CW
   $MAILSHARE/db/routes.bitnet contains:
	academic         bsmtp3!mailer!academic
	acmvm            bsmtp3!mailer!acmvm
	acusd            defrt1!acusd
	finhutc          bsmtp3nd!mailer!finhutc

   $MAILSHARE/sm.conf  contains:
	#
	# bitnet stuff F=hu not set
	#
	bsmtp3   snmSX /usr/local/lib/huji/bmail  bmail -b $h -u MAILER $g $u
	bsmtp3nd snmSX /usr/local/lib/huji/bmail  bmail -nd $h -u MAILER $g $u
	defrt1   snS   /usr/local/lib/huji/bmail  bmail $g $u
	bitnet2  snS   /usr/local/lib/huji/bmail  bmail $g $u

   $MAILSHARE/scheduler.cf  contains:
	defrt1/*
                command="sm -c $channel defrt1"
	bsmtp3/*
                command="sm -c $channel bsmtp3"
	bsmtp3nd/*
                command="sm -c $channel bsmtp3nd"
.ft R
.fi
.LP
Getting this installed into a BSD
.BR sendmail (8)
should not be very difficult, but definitely different..
Probably `mailer' definitions resemble  `sm.conf'  above.
.LP
.SH ENVIRONMENT
With environment variable
.B EBCDICTBL
user can select alternate EBCDIC <\-> ASCII translation
tables.  For more info, see
.BR ebcdictbl (5).
.LP
.SH FILES
.I /etc/funetnje.cf
.br
.I /etc/name.bitnet
(For system mailer)
.SH SEE ALSO
.BR bitspool (5),
.BR ebcdictbl (5),
.BR funetnje (8),
.BR mailify (1),
.BR qrdr (8),
.BR receive (1),
.BR sendfile (1),
.BR transfer (1)
.SH BUGS
.LP
None that we know.
.LP
.SH DIAGNOSTICS
These programs show their usage summary, when there is something
wrong with input parameters.  Subtle usage bugs might not be detected!
.SH WARNINGS
When BSMTP envelope address overflows to longer than 80-characters,
and NETDATA is not used, there is no check at it, nor folding...
(There is a deliberate TRUNCATE at 80 cols, though..)
With the NETDATA the input can be of arbitrary long lines (up to 64000
characters each).
.LP
.SH AUTHORS
.LP
Yehavi Bourvine, Hebrew University of Yerusalem, Israel
.I <yehavi@hujivms.bitnet>.
.LP
Matti Aarnio, Finnish University and Research Network \- FUNET, Finland
.I <mea@nic.funet.fi> <mea@finfiles.bitnet>.
.LP
Original (VMS oriented package) software by Yehavi Bourvine,
UNIX features, plenty of modifications, man-pages by Matti Aarnio.


ChangeLog
Fri Mar 22 13:48:30 1996  Matti Aarnio  <mea@nic.funet.fi>

	* libsendcmd.c, unix.c:
	    Even more relaxed approach on parametrizing
	    the CMDMAILBOX -- now allows SOCK/FIFO/UDP
	    with fairly flexible (in regards of spaces)
	    syntax..

Mon Nov 13 17:48:52 1995  Matti Aarnio  <mea@nic.funet.fi>

	* libsendcmd.c, unix.c:
	    A bit more relaxed approach on parametrizing
	    CMDMAILBOX config parameter (FIFO/SOCKET case)

	* file_queue.c:
	    Do RSCS-accounting on locally sent files, that
	    wind up to local recipients

	* unix_brcst.c:
	    Re-enter the SysVR4 UTMP weirdos..  Now fixed ?

Fri Aug 25 09:02:55 1995  Matti Aarnio  <mea@nic.funet.fi>

	* mailify.c:
	    Handling error-messages (where "MAIL FROM:" address
	    is "<>", that is, empty) was done improperly.  Now
	    it does flag input as from "channel error" in that
	    case.

Mon Jul 24 14:29:59 1995  Matti Aarnio  <mea@nic.funet.fi>

	* bmail.c, sendfile.c:
	    Sometimes   mktemp()  fails when there are a LOT of
	    files being sent at the same time.  Changed now to
	    more classical  ".MaiL.%05d"-format with  getpid()
	    result as integer data..

Wed Jun 21 22:42:39 1995  Matti Aarnio  <mea@nic.funet.fi>

	* unix_brcst.c, nmr_unix.c:
	    Ok, more than __linux__ is needed to know which
	    machines have <utmp.h> containing extended flags.
	    Now we test  "DEAD_PROCESS" pre-processor define,
	    which existence (propably) will correlate with
	    the existence of utmp's field  ut_type.
	    (SysVr4 feature, I think.)

Mon Jun 19 15:17:55 1995  Matti Aarnio  <mea@nic.funet.fi>

	* send.c, sendfile.c, bmail.c:
	    Some additional tests to trap various forms of bad
	    addresses -- accepts only  user@node format with
	    no '.' in the address, nor either component longer
	    than 8 chars.

	* nmr.c:
	    Made sure not to overwrite allocated string spaces on
	    Msg-queue entry.  (Did cause a crash with it..)

Tue Jun 13 09:13:18 1995  Matti Aarnio  <mea@nic.funet.fi>

	* qrdr.c:
	    Collect directory browsing results to an array at first,
	    then qsort() it with the time-stamp, and finally display
	    them in order. (Latest last)
	    (This is something what Gerald Hanush asked for.)

Mon Jun 12 22:44:00 1995  Matti Aarnio  <mea@nic.funet.fi>

	* file_queue.c:
	    Apparently OSF/1 pipe() return value is the first
	    fd, not "0", when it succeeds.  Nice thing - and not..
	    Altered the test to see if it returns a value under
	    zero, or something positive.  When the return is
	    positive (zero, or larger), fork() and run in
	    child-mode.

	    If a child_pipe exists already ( >= 0 ), close
	    it and mark the new one for processing.

	* libreadcfg.c, read_config, libsendcmd.c, unix.c:
	    Rewrote things so that now the used CMDMAILBOX
	    transport method is chosen at runtime:

	    CMDMAILBOX  F  /path/to/.cmdfifo
	    CMDMAILBOX  f  /path/to/.cmdfifo
	    CMDMAILBOX  S  /path/to/.cmdsocket
	    CMDMAILBOX  U  127.0.0.2 175	# UDP port

	    The "f" is FIFO which is closed, and reopened
	    every time one command is pulled in.

Sun Jun 11 03:12:06 1995  Matti Aarnio  <mea@nic.funet.fi>

	* namesparser.c:
	    Noticed that the data parsing stopped at first ":",
	    which caused premature breaks; observe following entry
	    from BITEARN.NODES -file:
	    	:remark1.BASE(9501) 00/00/00 00:00:00

	* libreadcfg.c, read_config.c, libsendcmd.c, unix.c:
	   Amended "CMDMAILBOX"-data to contain port-number in
	   case UDP-format is used. (That is, have two tokens.)
	   Default value for the port-number is 175.

	* unix_tcp.c read_passive_tcp_connection():
	   Noticed that a case when a socket exists, and a new
	   connection is attempted is handled improperly. The
	   old socket-fd was FORGOTTEN!  Now it NAKs the connection
	   attempt with a complaint that there already is a connection
	   in there...

	   Things will progress from such a situation usually by
	   detecting a closure on the existing stream, but better
	   not to rush ahead of things.

	* bmail.c:
	   Accept "-u" only when calling the program from uids 0 or 1
	   (root and daemon in most common cases -- for BSD-sendmail
	    which wants  bmail executable with protection 2111 (user X,
	    group NJEGRP) )

Sat Jun 10 05:18:27 1995  Matti Aarnio  <mea@nic.funet.fi>

	* Makefile, consts.h, prototypes.h, recv_file.c, io.c,
	  send_file.c, unix.c, protocol.c, unix_tcp.c:

	    - Found one uninitialized pointer, which apparently
	      caused mystic data-structure corruptions (on timer
	      subsystem)

	    - Made mandatory (well, with disabling options) to
	      use NBSTREAM, USE_XMIT_QUEUE, and USE_SOCKOPT.
	      (in consts.h, see also  Makefile's  comments!)

	    - Use  F_WAIT_V_A_BIT  to flag if there is any reason
	      for NOT to send out data at the moment, and instead
	      quit from within  handle_ack()..
	      The F_WAIT_V_A_BIT is set when new data is sent to
	      outbound buffer (primary one, or XMIT QUEUE).
	      The bit is cleared when pending writer finishes
	      with success status.  When the pending-writer is
	      called from  poll_sockets(), immediately after it
	      the  dequeue_xmit_queue() is called, and if there
	      is something, it can cause F_WAIT_V_A_BIT to be
	      set again..)

	    - If the XMIT_QUEUE ever runs out of space, it abort()s!

	    - Tested the beast by sending 80 or so smallish files both
	      ways over the link.  Such setup caused previously link
	      corruption..  Actually the critical phase is when there
	      are flowing NJE headers/trailer, bare data sending works
	      just fine, but when a file ends, EOF block needs an ack
	      to be sent, and unless there the care is used, all
	      queues will overflow..

Thu Jun  8 15:48:14 1995  Matti Aarnio  <mea@nic.funet.fi>

	* namesfilter.c:
	    New option: "-netsoft" to list node software fields.

	* funetnje.texinfo:
	    Some additional chapters

	* unix_tcp.c:

bmail.c
/*   BMAIL.C	V1.2 - mea-1.4
 | Copyright (c) 1988,1989,1990 by
 | The Hebrew University of Jerusalem, Computation Center.
 |
 | Copyright (c) 1991, 1993 by
 | Finnish University and Research Network, FUNET
 |
 |   This software is distributed under a license from the Hebrew University
 | of Jerusalem. It may be copied only under the terms listed in the license
 | agreement. This copyright message should never be changed or removed.
 |   This software is given without any warranty, and the Hebrew University
 | of Jerusalem assumes no responsibility for any damage that might be caused
 | by use or misuse of this software.
 |
 | Calling sequence:  BMAIL from to LinkName
 | or, for BSMTP: BMAIL from to LinkName -b GateWay OurName
 |
 | V1.1 - 22/3/90 - Upcase From, To and LinkName as lower case might cause
 |        problems to IBMs.
 | V1.2 - 4/4/90 - Correct the Usage note.
 | V1.3 - 10-Sep-93 - Some transmission finesses..
 | V1.4 - 10-Oct-93 - Teached it to transmit NETDATA
 */


#include "prototypes.h"
#include "clientutils.h"
#include "ndlib.h"
#include <sysexits.h>

#ifndef MAILERNAME	/* Name of the user that "SENDS" email */
#define MAILERNAME "MAILER"
#endif

#define	LSIZE	256			/* Maximum line length */

char LOCAL_NAME   [10];
char BITNET_QUEUE [80];
char LOG_FILE     [80] = "-";	/* STDERR as the default */
char mailername   [10];

int LogLevel = 1;
FILE *LogFd = NULL;

unsigned char linehdr[2] = { 0x80, 80 };

void
usage(str)
char *str;
{
  fprintf(stderr, "Usage: bmail [-u origuser] [-v] [-tag RSCS-TAG] [-M mailername]\n             [-nd Gateway-address | -b(smtp) Gateway-address]\n             FromUser@FromNode ToUser@ToNode [ToUser@ToNode...]\n");
  fprintf(stderr, "  File is taken from the standard input\n");
  if (str)
    fprintf(stderr,"ErrMsg: %s\n",str);
  exit(1);
}

int
main(argc, argv)
int	argc;
char	*argv[];
{
	char	ebcdicline[LSIZE]; /* 80 needed, more available.. */
	char	outbuf[85];
	char	*From = NULL, FileName[LSIZE], tFileName[LSIZE],
		*GateWay = NULL, To[LSIZE];
	char	line[1000], *p, *q;
	char	NjeFrom[20];
	FILE	*fd;
	int	Netdata = 0, Bsmtp = 0, ForcedMailer = 0;
	int	FileSize, i, rc;
	struct stat stats;
	char	*Tag = NULL;
	long	reccount = 0;
	long	recpos = 0;
	int	verbose = 0;
	int	head = 1; /* Clear, when past the headers. P Bryant */
	char	*origuser = NULL;
	int	myuid = getuid();

	memset( ebcdicline, 0, 80 );
	strncpy(mailername,MAILERNAME,8);
	mailername[8] = 0;



	if (argc < 3) usage("Too few arguments");

/* Read our name and the BITnet queue */
	read_configuration();
	read_etable();

	++argv;
	while (*argv) {
	  if (strcmp(*argv, "-tag") == 0) {
	    ++argv;
	    if (!*argv) usage("-tag missing mandatory data");
	    Tag = *argv;
	  } else if (strcmp(*argv, "-nd") == 0) {
	    Bsmtp = 1;
	    Netdata = 1;
	    ++argv;
	    if (!*argv) usage("-nd (bsmtp) missing mandatory data");
	    GateWay = *argv;
	    upperstr(GateWay);
	  } else if (strcmp(*argv, "-u") == 0) {
	    ++argv;
	    if (!*argv) usage("-u missing mandatory data");
	    origuser = *argv;
	    upperstr(origuser);
	    if (!(myuid == 0 || myuid == 1)) {
	      /* FIXME: HARDWIRED ROOT AND DAEMON!
	                Daemon is usually the uid which BSD-sendmail
			uses when spawning a child to do transportation.. */
	      usage("-u can be used by 'root' and 'daemon' users only!");
	    }
	  } else if (strcmp(*argv, "-b") == 0 ||
		     strcmp(*argv, "-bsmtp") == 0) {
	    Bsmtp = 1;
	    ++argv;
	    if (!*argv) usage("-b(smtp) missing mandatory data");
	    GateWay = *argv;
	    upperstr(GateWay);
	  } else if (strcmp(*argv,"-M") == 0) {
	    ++argv;
	    if (!*argv) usage("-M missing mandatory data");
	    strncpy(mailername, *argv, 8);
	    mailername[8] = 0;
	  } else if (strcmp(*argv,"-v") == 0) {
	    verbose = 1;
	  } else break;
	  ++argv;
	}
	From  = *argv++;
	if (!From || !*argv) usage("No From, and/or To missing");
	if (!Bsmtp && argv[1]) usage("No BSMTP, and multiple recipients");

/* Upcase linkname, From and To to prevent from problems with RSCS */
	upperstr(From);

	/* If trailed with ".BITNET", chop it.. */
#define CHOP_BITNET(VarName) \
	i = strlen((VarName)); \
	if (i > 8 && memcmp((VarName)+i-7,".BITNET",7) == 0) \
	  (VarName)[i-7] = 0
	CHOP_BITNET(From);

	strncpy(To,*argv,sizeof(To)-1);
	upperstr(To);
	CHOP_BITNET(To);

	/* A SGID program.. */
	setgid(getegid());

	sprintf(tFileName, "%s/.MaiL%05d", BITNET_QUEUE,getpid());
	if (*tFileName == '\0') {
		fprintf(stderr, "Can't create unique filename\n");
		exit(1);
	}

	umask(077);
	if ((fd = fopen(tFileName, "w")) == NULL) {
		perror(tFileName); exit(1);
	}
	if (fstat(fileno(fd),&stats)) {
	  perror("Can't fstat(2) ??"); unlink(tFileName); exit(2);
	}

	fprintf(fd, "%-506s\nEND:\n","*");
	fflush(fd);
	fseek(fd,0,0);

	p = (char *)strchr(From,'@');
	if (!p || p > &From[8]) ForcedMailer = 1;
	if (p && (q = (char *)strchr(p,'.'))) ForcedMailer = 1;

/* The RSCS envelope for our NJE emulator */
	if (!origuser && (ForcedMailer || Bsmtp))
	  origuser = mailername;
	if (Bsmtp || ForcedMailer || origuser)
	  sprintf(NjeFrom, "%.8s@%.8s", origuser, LOCAL_NAME);
	else
	  strncpy(NjeFrom, From, 19);
	NjeFrom[20] = 0;
	fprintf(fd, "FRM: %-17.17s\n", NjeFrom);
	fprintf(fd, "FMT: BINARY\nTYP: MAIL\nEXT: MAIL\nCLS: M\nFOR: QUMAIL\n");
	if (Bsmtp || ForcedMailer)
	  fprintf(fd, "FNM: %-12s\n", mailername);	/* The `username' that sends it */
	else
	  fprintf(fd, "FNM: %-12s\n", mcuserid(0));	/* The username that sends it */
	if(Bsmtp)
		fprintf(fd, "TOA: %-17.17s\n", GateWay);
	else
		fprintf(fd, "TOA: %-17.17s\n", To);
	fprintf(fd, "FID: 0000\nOID: 0000\n");
	if (Tag)
	  fprintf(fd, "TAG: %-136.136s\n",Tag);
	recpos = ftell(fd);
	fprintf(fd, "REC: %8d",1); /* NO \n ! */

	fflush(fd);

	fseek(fd,0,2); /* To the end.. */

	if (Netdata) {

	  /* Netdata mode implies BSMTP, we figure out the lrecl.. */

	  int	c, c1;
	  int	lrecl = 80;	/* Lets assume at least this `little' */
	  int	lencnt;
	  struct puncher PUNCH;

	  /* We need an ASCII temporary file.. */
	  FILE *buffile = tmpfile();
	  
	  /* Prepare a puncher.. */

	  PUNCH.buf[0] = CC_NO_SRCB;
	  PUNCH.buf[1] = 80;
	  PUNCH.fd       = fd;
	  PUNCH.len      = 0;
	  PUNCH.punchcnt = 0;

	  /* -- Prepare BSMTP envelope file -- */

	  sprintf(line,"HELO %s.BITNET\n", LOCAL_NAME);
	  lrecl = strlen(line)-1;
	  fputs(line,buffile);
	  if (verbose)
	    fprintf(buffile,"VERB ON\n");
	  fprintf(buffile,"TICK %d\n",(int)time(NULL));
	  sprintf(line,"MAIL FROM:<%s>\n", From);
	  fputs(line,buffile);
	  lencnt = strlen(line)-1;
	  if (lencnt > lrecl) lrecl = lencnt;
	  while (*argv) {
	    sprintf(line,"RCPT TO:<%s>\n", *argv);
	    fputs(line,buffile);
	    lencnt = strlen(line)-1;
	    if (lencnt > lrecl) lrecl = lencnt;
	    ++argv;
	  }
	  fprintf(buffile,"DATA\n");
	  c1 = 0;

	  /* Copy the file over, find longest line's length.. */
	  while ((c = getc(stdin)) != EOF) {
	    putc(c,buffile);

	    /* If our write returns a failure, exit! */
	    if (ferror(buffile)) {
	      fprintf(stderr,"BMAIL/ND: write of BSMTP temp file failed!\n");
	      unlink(tFileName);
	      exit(EX_TEMPFAIL);
	    }

	    if (lencnt == 0 && c == '.')
	      putc(c,buffile);	/* Duplicate front '.' */
	    if (c == '\n') {
	      if (lencnt > lrecl) lrecl = lencnt;
	      lencnt = -1;
	    }
	    ++lencnt;
	    c1 = c;
	  }
	  if (c1 != '\n')
	    putc('\n',buffile);
	  fprintf(buffile,".\nQUIT\n");
	  fseek(buffile,0,0);	/* rewind() ! */

	  /* -- And output it.. -- */
	  if (lrecl < 79) lrecl = 79;
	  if (!do_netdata(buffile,&PUNCH,NjeFrom,To,
			  mailername,"MAIL",lrecl+1,ND_VARY,0,NULL)) {
	    fprintf(stderr,"BMAIL/ND: write of bitnet spool file failed!\n");
	    unlink(tFileName);
	    exit (EX_TEMPFAIL);
	  }

	  fclose(buffile);

	  reccount += PUNCH.punchcnt;

	} else {

#define eprintf(x,y) { 					\
	  sprintf(ebcdicline,x,y);  ebcdicline[80] = 0;	\
	  i = strlen(ebcdicline);			\
	  memcpy( outbuf+0,linehdr,2 );			\
	  ASCII_TO_EBCDIC( ebcdicline,outbuf+2,i );	\
	  if (!Uwrite( fd,outbuf,2+i )) failure = 1;	\
	  ++reccount;  }

	  int failure = 0;

	  if (Bsmtp) {
	    /* If this is a BSMTP message, add the BSMTP header: */
	    eprintf ("HELO %s.BITNET", LOCAL_NAME);
	    if (verbose)
	      eprintf("VERB %s","ON");
	    eprintf ("TICK %d",(int)time(NULL)); /* Not used actually... */
			/* WARNING: MAY FOLD OVER... NEED MORE CODE HERE... */
	    eprintf ("MAIL FROM:<%s>", From);
	    while (*argv) {
	      eprintf ("RCPT TO:<%s>", *argv);
	      ++argv;
	    }
	    eprintf ("%s","DATA");
	  }


	  /* The message itself - copy it as-is */
	  while (!failure && fgets(line+1, sizeof(line)-1, stdin) != NULL) {

	    /* If it is BSMTP, duplicate the dots at
	       the beginning of a line */

	    int len = strlen(line+1);
	    char *l;

	    p = line + len +1;
	    if (len>0)		/* Zap the \n! */
	      if (*(p-1)=='\n')
		*(--p) = 0;
	    l = line+1;
	    if (--len == 0) head = 0;	/* Blank line indicated end of header */
	    while (!failure) {	/* Fold long input.. */
	      int tempchar;
	      if (Bsmtp && (*l == '.')) {
		/* Duplicate the (sub)line starting `.' */
		--l;
		*l = '.';
		++len;
	      }
	      /* P Bryant 10/1/95 code to sort out folding */
	      p=l+len;
	      if (len > 80) {	/*must fold in header*/
		/* now serch back for separator */
		p=l+80;
		if (head != 0) while (*p != ' ' && p != l ) --p;
		/* if p=l we cannot fold so hope for best */
		if (p == l) p=l+80;
	      }
	      tempchar = *p;
	      *p = 0;		/* terminate record */
	      eprintf("%-80.80s",l);
	      *p = tempchar;
	      len = len -( p - l);
	      if (len <= 0) break;
	      l = p;
	      if (*l != ' ' && head != 0){ *(--l)= ' ';++len;}
	      /*	      eprintf("%-80.80s",l);
			      l   += 80;
			      len -= 80; */
	    }
	  } /* While fgets() -loop */

	  /* If BSMTP - add the .QUIT: */

	  if (!failure && Bsmtp != 0) {
	    eprintf("%s",".");
	    eprintf("%s","QUIT");
	  }
	  if (failure) {
	    fprintf(stderr,"BMAIL/%s: bitnet spool write failed!\n",
		    Bsmtp ? "BSMTP":"PUNCH");
	    unlink(tFileName);
	    exit (EX_TEMPFAIL);
	  }
	} /* End of !Netdata */
	
	FileSize = ftell(fd);	/* Get its size in bytes */

	fseek(fd, recpos, 0);
	fprintf(fd, "REC: %8ld",reccount);

	fclose(fd);

	sprintf(FileName, "%s/MaiL%ld", BITNET_QUEUE,(long)stats.st_ino);
	rename(tFileName,FileName);

	/* Queue it to the NJE emulator */
	
	rc = submit_file(FileName, FileSize);
	if (rc)
	  unlink(FileName);
	exit (rc);
}
